### PR TITLE
1001 raster unitest

### PR DESCRIPTION
**Pull request by gantian127**
This pull request needs to be merged after PR #989 is merged.

### CODE REVIEW

**Comment by hydrobuild** [comment]
Can one of the admins verify this patch?

File: hs_geo_raster_resource/receivers.py [hunk 1]
```diff
@@ -1,48 +1,171 @@
...
 @receiver(pre_create_resource, sender=RasterResource)
 def raster_pre_create_resource_trigger(sender, **kwargs):
     files = kwargs['files']
     title = kwargs['title']
-
+    validate_files_dict = kwargs['validate_files']
     metadata = kwargs['metadata']
-    from collections import OrderedDict
+
     if(files):
```
Comment (pkdash) [inline hs_geo_raster_resource/receivers.py]: parentheses not needed

File: hs_geo_raster_resource/receivers.py [hunk 2]
```diff
@@ -140,6 +278,17 @@ def raster_pre_delete_file_from_resource_trigger(sender, **kwargs):
     res.metadata.create_element('BandInformation', name='Band_1', variableName='Unnamed', variableUnit='Unnamed',
                                 method='', comment='')
 
+    # delete all the files that is not the user selected file
+    for f in ResourceFile.objects.filter(object_id=res.id):
+        if f.resource_file.name != del_file.resource_file.name:
+            f.resource_file.delete()
```
Comment (pkdash) [inline hs_geo_raster_resource/receivers.py]: Better to use the ```delete_resource_file()``` available in hs_core\hydroshare\resource.py

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+
+        # trying to add good .tif file should pass the file check
+        files = [UploadedFile(file=self.raster_tif_file_obj, name=self.raster_tif_file_name)]
+        utils.resource_file_add_pre_process(resource=self.resRaster, files=files, user=self.user,
+                                            extract_metadata=False)
+        utils.resource_file_add_process(resource=self.resRaster, files=files, user=self.user,
+                                        extract_metadata=False)
+
+        # there should be 2 content file: with .vrt file created by system
+        self.assertEquals(self.resRaster.files.all().count(), 2)
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: May be you need to also test that a .vrt file exists as part of the resource.

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        # delete content file that we added above
+        hydroshare.delete_resource_file(self.resRaster.short_id, self.raster_tif_file_name, self.user)
+
+        # trying to add good .zip file should pass the file check
+        files = [UploadedFile(file=self.raster_zip_file_obj, name=self.raster_zip_file_name)]
+        utils.resource_file_add_pre_process(resource=self.resRaster, files=files, user=self.user,
+                                            extract_metadata=False)
+        utils.resource_file_add_process(resource=self.resRaster, files=files, user=self.user,
+                                        extract_metadata=False)
+
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: Should test here that the resource has the added file.

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+                                            extract_metadata=False)
+
+        utils.resource_file_add_process(resource=self.resRaster, files=files, user=self.user,
+                                        extract_metadata=True)
+        # there should be 2 content files
+        self.assertEquals(self.resRaster.files.all().count(), 2)
+
+        # there should be 2 format elements
+        self.assertEquals(self.resRaster.metadata.formats.all().count(), 2)
+
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: Should check expected format values also.

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+        files = [UploadedFile(file=self.raster_tif_file_obj, name=self.raster_tif_file_name)]
+        utils.resource_file_add_pre_process(resource=self.resRaster, files=files, user=self.user,
+                                            extract_metadata=False)
+
+        utils.resource_file_add_process(resource=self.resRaster, files=files, user=self.user,
+                                        extract_metadata=True)
+
+        # before resource delete
+        # resource core metadata
+        core_metadata_obj = self.resRaster.metadata
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: This is not the 'core metadata'. It includes extended metadata too.

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+        self.assertEquals(CoreMetaData.objects.all().count(), 1)
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: It should be ```RasterMetaData``` and not ```CoreMetaData```

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+        # there should be original coverage metadata objects
+        self.assertTrue(OriginalCoverage.objects.filter(object_id=core_metadata_obj.id).exists())
+        # there should be CellInformation metadata objects
+        self.assertTrue(CellInformation.objects.filter(object_id=core_metadata_obj.id).exists())
+        # there should be BandInformation metadata objects
+        self.assertTrue(BandInformation.objects.filter(object_id=core_metadata_obj.id).exists())
+
+        # delete resource
+        hydroshare.delete_resource(self.resRaster.short_id)
+        self.assertEquals(CoreMetaData.objects.all().count(), 0)
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: Same as above.

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+            'user1@nowhere.com',
+            username='user1',
+            first_name='Creator_FirstName',
+            last_name='Creator_LastName',
+            superuser=False,
+            groups=[self.group]
+        )
+
+        # even there is no file uploaded to resource, there are default extended automatically metadata created
+        _, _, metadata = utils.resource_pre_create_actions(resource_type='RasterResource',
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: Should test what ```metadata``` is created by ```resource_pre_create_actions()```.

File: hs_geo_raster_resource/receivers.py (hunk 2 above)
Comment (gantian127) [inline hs_geo_raster_resource/receivers.py]: delete_resource_file() function send signal of pre_delete_file_from_resource(). It seems these two functions will call each other as a loop if I use delete_resource_file() in this signal function.

**Comment by gantian127** [comment]
@pkdash this is ready for merge.

File: hs_geo_raster_resource/receivers.py [hunk 10]
```diff
@@ -308,7 +315,7 @@ def metadata_element_pre_update_handler(sender, **kwargs):
         element_form = CellInfoValidationForm(request.POST)
     elif element_name == 'bandinformation':
         form_data = {}
-        for field_name in BandInfoValidationForm().fields:
+        for field_name in BandInfoValidationForm().fsields:
```
Comment (pkdash) [inline hs_geo_raster_resource/receivers.py]: Are you sure the change to ```fsields``` is correct?

File: hs_geo_raster_resource/tests/test_raster_metadata.py [hunk 11]
```diff
@@ -0,0 +1,569 @@
...
+                                        extract_metadata=False)
+
+        # there should be 2 content file: with .vrt file created by system
+        self.assertEquals(self.resRaster.files.all().count(), 2)
+        file_names = [os.path.basename(f.resource_file.name) for f in self.resRaster.files.all()]
+        self.assertIn('raster_tif_valid.vrt', file_names)
+
+        # delete content file that we added above
+        hydroshare.delete_resource_file(self.resRaster.short_id, self.raster_tif_file_name, self.user)
+
```
Comment (pkdash) [inline hs_geo_raster_resource/tests/test_raster_metadata.py]: How many content files the resource should have at this point? Should test that before adding the zip file.

File: hs_geo_raster_resource/receivers.py (hunk 10 above)
Comment (gantian127) [inline hs_geo_raster_resource/receivers.py]: that is a typo. Thanks for finding it out.

**Comment by pkdash** [comment]
@gantian127 I have couple of additional comments. Once you address those I can +1.

**Comment by gantian127** [comment]
@pkdash please review the code update and let me know if I need to make more edits. Thanks~

**Comment by pkdash** [comment]
+1

**Pull request merged**